subsys/bluetooth/services/ias/ias_client.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdint.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/services/ias.h>
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_IAS_CLIENT)
+#define LOG_MODULE_NAME bt_ias_client
+#include "../../common/log.h"
+
+enum {
+	IAS_DISCOVER_IN_PROGRESS,
+
+	IAS_NUM_FLAGS, /* keep as last */
+};
+
+struct bt_ias_client {
+	/* Handle for alert writes */
+	uint16_t alert_level_handle;
+
+	/** Internal flags **/
+	ATOMIC_DEFINE(flags, IAS_NUM_FLAGS);
+
+	/* Gatt discover procedure parameters */
+	struct bt_gatt_discover_params discover;
+};
+
+static const struct bt_uuid *alert_lvl_uuid = BT_UUID_ALERT_LEVEL;
+static const struct bt_uuid *ias_uuid = BT_UUID_IAS;
+
+static const struct bt_ias_client_cb *ias_client_cb;
+static struct bt_ias_client client_list[CONFIG_BT_MAX_CONN];
+
+static struct bt_ias_client *client_by_conn(struct bt_conn *conn)
+{
+	return &client_list[bt_conn_index(conn)];
+}
+
+static void client_cleanup(struct bt_ias_client *ias_client)
+{
+	(void)memset(ias_client, 0, sizeof(*ias_client));
+}
+
+static void discover_complete(struct bt_conn *conn, int err)
+{
+	BT_DBG("conn %p", (void *)conn);
+
+	if (err) {
+		client_cleanup(client_by_conn(conn));
+		BT_DBG("Discover failed (err %d\n)", err);
+	}
+
+	if (ias_client_cb != NULL && ias_client_cb->discover != NULL) {
+		ias_client_cb->discover(conn, err);
+	}
+}
+
+int bt_ias_client_alert_write(struct bt_conn *conn, enum bt_ias_alert_lvl lvl)
+{
+	int err;
+
+	CHECKIF(conn == NULL) {
+		return -ENOTCONN;
+	}
+
+	if (client_by_conn(conn)->alert_level_handle == 0) {
+		return -EINVAL;
+	}
+
+	err = bt_gatt_write_without_response(conn,
+					     client_by_conn(conn)->alert_level_handle,
+					     &lvl, sizeof(lvl), false);
```

Comment:
Given that `lvl` is an `enum`, we should ensure that the size and values are correct before sending it over the air. 

I assume it's an 8-bit value, right? 

Generally we shouldn't assume anything when doing `sizeof(<enum>)`, as `enum` is not a fixed width type.

Reply:
Yup, it's an 8-bit value, I've added a check that returns an invalid argument error if `lvl` size differs from `uint8_t` size.

Reply:
This is a poor solution, because if `sizeof(lvl) != sizeof(uint8_t)` then the user has no choice but to use a different compiler. 

A better solution is to create another local variable as an `uint8_t`, assign the `lvl` value to the `uint8_t` (assuming that the `lvl` value is correct) and then use the `sizeof` on the newly create value. 

See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/audio/pacs.c#L267-L295 on how PACS converts an `enum` to a fixed size value.